### PR TITLE
[SPARK-46488][SQL] Skipping trimAll call during timestamp parsing

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?

This PR is a follow up to [46173](https://github.com/apache/spark/pull/44110) which added skipping the trimAll calls during date parsing.

Now I'm doing the same just for timestamp parsing.

### Why are the changes needed?

These changes should drastically improve edge case where input string in cast to date has many whitespace as prefix/sufix.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

I added the tests to check for cases with prefixes and suffixes of whitespaces and control chars.

Also there are benchmark tests in the previous [PR](https://github.com/apache/spark/pull/44110)

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dbatomic** [comment]
LGTM

File: sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala [hunk 1]
```diff
@@ -619,6 +616,29 @@ trait SparkDateTimeUtils {
...
+
+    while (start < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(start))) {
+      start += 1
+    }
+
+    while (end > start && UTF8String.isWhitespaceOrISOControl(bytes(end))) {
+      end -= 1
+    }
+
+    (start, end + 1)
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala:640]: Don't you create a `Tuple` instance here. Is it possible to avoid this? For example, define two separate `inline` functions:
- `getTrimmedStart(bytes: Array[Byte]): Int`
- `getTrimmedEnd(bytes: Array[Byte], start: Int): Int`

Comment (stefankandic) [reply sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala]: done :)

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @stefankandic and @dbatomic for review.

**Comment by MaxGekk** [comment]
@stefankandic Congratulations for your first contribution to Apache Spark!
